src/coprocessor/local_metrics.rs

```diff
@@ -11,46 +11,151 @@
+        CopFlowStatistics {
+            sender: sender,
+            data: Default::default(),
+        }
+    }
+
+    pub fn add(&mut self, region_id: u64, stats: &Statistics) {
+        let flow_stats = self.data.entry(region_id).or_default();
+        flow_stats.add(&stats.write.flow_stats);
+        flow_stats.add(&stats.data.flow_stats);
+    }
+
+    pub fn flush(&mut self) {
+        if self.data.is_empty() {
+            return;
+        }
+        let mut to_send_stats = HashMap::default();
+        mem::swap(&mut to_send_stats, &mut self.data);
+        if let Err(e) = self.sender.schedule(PdTask::ReadStats {
+            read_stats: to_send_stats,
+        }) {
+            error!("send coprocessor statistics: {:?}", e);
+        };
     }
+}
+
+/// `ExecLocalMetrics` collects metrics from `ExectorMetrics`
+pub struct ExecLocalMetrics {
+    flow_stats: CopFlowStatistics,
+    scan_details: LocalCounterVec,
+    scan_counter: LocalCounterVec,
+    exec_counter: LocalCounterVec,
+}
 
-    #[inline]
-    pub fn inc_point(&mut self) {
-        self.point += 1;
+impl ExecLocalMetrics {
+    pub fn new(sender: FutureScheduler<PdTask>) -> ExecLocalMetrics {
+        ExecLocalMetrics {
+            flow_stats: CopFlowStatistics::new(sender),
+            scan_details: COPR_SCAN_DETAILS.local(),
+            scan_counter: COPR_GET_OR_SCAN_COUNT.local(),
+            exec_counter: COPR_EXECUTOR_COUNT.local(),
+        }
     }
 
-    /// Merge records from `other` into `self`, and clear `other`.
-    #[inline]
-    pub fn merge(&mut self, other: &mut ScanCounter) {
-        self.range += other.range;
-        self.point += other.point;
-        other.range = 0;
-        other.point = 0;
+    pub fn collect(&mut self, type_str: &'static str, region_id: u64, metrics: ExecutorMetrics) {
+        let stats = &metrics.cf_stats;
+        // cf statistics group by type
+        for (cf, details) in stats.details() {
+            for (tag, count) in details {
+                self.scan_details
+                    .with_label_values(&[type_str, cf, tag])
+                    .inc_by(count as f64)
+                    .unwrap();
+            }
+        }
+        // flow statistics group by region
+        self.flow_stats.add(region_id, stats);
+        // scan count
+        metrics.scan_counter.consume(&mut self.scan_counter);
+        // exec count
+        metrics.executor_count.consume(&mut self.exec_counter);
     }
 
-    #[inline]
     pub fn flush(&mut self) {
-        if self.range > 0 {
-            let range_counter = COPR_GET_OR_SCAN_COUNT.with_label_values(&["range"]);
-            range_counter.inc_by(self.range as f64).unwrap();
-            self.range = 0;
+        self.flow_stats.flush();
+        self.scan_details.flush();
+        self.scan_counter.flush();
+        self.exec_counter.flush();
+    }
+}
+
+/// `BasicLocalMetrics` is for the basic metrics for coprocessor requests.
+pub struct BasicLocalMetrics {
```

Comment:
I feel it strange to have two metrics containers here, `BasicLocalMetrics` and `ExecLocalMetrics`. What kind of metrics should be placed in `BasicLocalMetrics` and what should be placed in another? If we are adding more metrics in the future, what's the rule to determine where it should be placed into? Can you explain necessary reasons to separate metrics into two places?

Reply:
BasicLocalMetrics are metrics for all request, whether it has been failed in params check or not.
ExecLocalMetrics are metrics generated by executors. For example, in the future we may support Key-Value command in the coprocessor, it may have BasicLocalMetrics, while has no ExecLocalMetrics. @breeswish

Reply:
@AndreMouche got it